### ISSUE TITLE
Implement end to end push tests for Cordova

### ISSUE THREAD

**Issue by shrishrirang** [Azure/azure-mobile-apps-js-client#63]
- The push tests in the Cordova end to end test suite test the push Cordova plugin.
- There is no coverage for the Mobile Apps Push registration APIs
- Implement tests for testing Mobile Apps push registration APIs.

**Comment by samiroquai** [comment]
Hi, I've installed the beta yesterday (https://azure.microsoft.com/en-us/blog/announcing-the-beta-apache-cordova-plugin-for-azure-app-service-mobile-apps/)

Should the push.register method work? I'd like to know before spending too much time trying to find what's wrong with my test here :)

Thank you.

**Comment by shrishrirang** [comment]
Yes, the beta 3 release of the Cordova plugin supports push registration for gcm, apns and wns.

What is the failure you are seeing and on which platform? You need to use the new push plugin https://github.com/phonegap/phonegap-plugin-push to obtain the registrationId which you can then use for registration.

Here are some useful code snippets to give you an overview:

Initialize push:

    pushNotification = PushNotification.init({
                        android: {
                            <<senderID>> // ***** Get this ID from Google console ******
                        },
                        ios: {
                            alert: "true",
                            badge: "true",
                            sound: "true"
                        },
                        windows: {}
                    });

Obtain the push registration ID:

    pushNotification.on('registration', function (data) {
        registrationId = data.registrationId;
    });

Processing notifications:

    pushNotification.on('notification', function (data) {
        window.alert('Received push notification : ' + JSON.stringify(data));
    });

Error handling:

    pushNotification.on('error', function (error) {
        window.alert('Error : ' + error);
    });

Register for push using the registration ID obtained above:

    client.push.register(<<name>>, registrationId); // name can be 'gcm', 'wns', 'apns'

Server can send push notifications using something like this:

    req.azureMobile.push.gcm.send(undefined /* tags */, '{"data":{"message":"android push notification"}}');
    req.azureMobile.push.apns.send(undefined /* tags */, '{"aps":{"alert":"iOS push notification"}}');
    req.azureMobile.push.wns.send(undefined /* tags */, '<?xml version="1.0"?><toast><visual><binding template="ToastText01"><text id="1">windows push notification</text></binding></visual></toast>', 'wns/toast');

**Comment by samiroquai** [comment]
Thank you for the code samples, it helped. I already did all of the above, except that I was trying to register with a template. 
I've followed your samples and I now use the "short" form of the register function (I guess a kind of native registration?) and I got a success.
I've then tried the integration of it all by sending a test message from Azure Portal. I got the notification on my device. 
So I can say it works.
Thank you for your time.

**Comment by shrishrirang** [comment]
:+1:

**Comment by phvannor** [comment]
Can you retry with templates and let us know if they are working?  Both should be valid, maybe include the template you are sending so we can verify its ok?

**Comment by samiroquai** [comment]
Sure I’ll try that tonight

**Comment by samiroquai** [comment]
Ok, I did the test again with template registration. I can't get it to work, my call is probably wrong. Here's what it looks like. 
```javascript
    push.on('registration', function(data) {
    alert("registered! "+ data.registrationId);
    var azureConnection = new WindowsAzure.MobileServiceClient('https://mysite.azurewebsites.net');
    azureConnection
        .push
        .register("apns",
                data.registrationId
                ,[{"body":{"aps":{"alert":"$(message)"}},
                   "headers":[], "tags":["General"]
                  }]
                ,null)
        .then(function(data){
                alert("success");
            },function(error){
                alert(error);
            });
    this.registrationId=data.registrationId;
});
```
On the other hand, here's the code that works:
```javascript
    push.on('registration', function(data) {
    alert("registered! "+ data.registrationId);
    var azureConnection = new WindowsAzure.MobileServiceClient('https://mysite.azurewebsites.net');
    azureConnection
        .push
        .register("apns",
                data.registrationId)
        .then(function(data){
                alert("success");
            },function(error){
                alert(error);
            });
    this.registrationId=data.registrationId;
});
```

**Comment by phvannor** [comment]
ok, the templates param is not an array, that is probably the issue.  Try passing,

{ "mytemplate" : { "body": { "aps" : { "alert" : "$(message)" } }, "headers" : [], "tags" : [ "general" ] } }
see: https://msdn.microsoft.com/en-us/library/azure/mt621153.aspx

Also, note that currently the "tags" property of each template will be ignored, we are working on adding a whitelist & programming hook server side to help control tags, but that is not yet complete.  Adding tags for now must be done on teh server (most likely via a custom API)

**Comment by samiroquai** [comment]
I tested with the parameter you proposed. Ends with a 500 (Internal Server Error). No more details.
Here's the code tested (I've tried with and without tags, no success).

```javascript
push.on('registration', function(data) {
    alert("registered! "+ data.registrationId);
    console.log("registered with APNS");
       var azureConnection = new WindowsAzure.MobileServiceClient('https://cerbere.azurewebsites.net');
            alert(azureConnection.push);
            azureConnection.push.register("apns",
                data.registrationId,
                { "mytemplate" : { "body": { "aps" : { "alert" : "$(message)" } }, "headers" : [], "tags" : [] } })
           .then(function(data){
                alert("success");
            },function(error){
                alert(error);
            });
    this.registrationId=data.registrationId;
});
```

**Comment by fiyazbinhasan** [comment]
Hi, I'm using the azure mobile apps js client recently and trying to enable push notification in one of my app. I've downloaded the cordova push plugin to generate registrationId for wns. The device is successfully registered for wns but when I try to send toast messages from VS push notification debug windows it says "The Token obtained from the Token Provider is wrong". I would I resolve this. I also tried from real windows device but the problem still exists.
![error](https://cloud.githubusercontent.com/assets/6568968/12534636/a8dc7f14-c28d-11e5-911f-1a807376ebb2.png)
